tests/base_new_frmwrk.py

```diff
@@ -150,6 +150,217 @@ def set_replication_methods(self, conn_id, catalogs, replication_methods):
+                    'frequency',
+                    'conversion_values',
+                    'canvas_avg_view_percent',
+                    'cost_per_conversion',
+                    'cost_per_thruplay',
+                    'action_values',
+                    'full_view_impressions',
+                    'place_page_name',
+                    'instant_experience_outbound_clicks',
+                    'cost_per_unique_action_type',
+                    'estimated_ad_recallers',
+                    'outbound_clicks',
+                    'social_spend',
+                    'cost_per_unique_inline_link_click',
+                    'instant_experience_clicks_to_start',
+                    'attribution_setting',
+                    'engagement_rate_ranking',
+                    'purchase_roas',
+                    'reach',
+                    'cost_per_outbound_click',
+                    'estimated_ad_recall_rate',
+                    'quality_ranking',
+                    'cpp',
+                    'catalog_segment_value',
+                    'canvas_avg_view_time',
+                    'cost_per_action_type',
+                    'outbound_clicks_ctr',
+                    'qualifying_question_qualify_answer_rate',
+                    'converted_product_quantity',
+                    'converted_product_value',
+                    'instant_experience_clicks_to_open',
+                    'conversion_rate_ranking'
+             },
+            "ads": {
+                    'bid_amount',
+                    'bid_info',
+                    'recommendations',
+                    'adlabels'
+            },
+            "ads_insights_platform_and_device": {
+                    'video_p75_watched_actions',
+                    'conversions',
+                    'conversion_values',
+                    'canvas_avg_view_percent',
+                    'action_values',
+                    'unique_outbound_clicks',
+                    'cost_per_unique_action_type',
+                    'outbound_clicks',
+                    'social_spend',
+                    'video_p50_watched_actions',
+                    'engagement_rate_ranking',
+                    'video_p25_watched_actions',
+                    'quality_ranking',
+                    'video_play_curve_actions',
+                    'conversion_rate_ranking',
+                    'video_30_sec_watched_actions',
+                    'canvas_avg_view_time',
+                    'cost_per_action_type',
+                    'video_p100_watched_actions',
+                    'placement'
+            }
+        }
+        actual_expected = {}
+        non_upsert_streams = non_upsert_fields.keys()
+        for key, fields in selected_fields.items():
+            if key in non_upsert_streams:
+                actual_expected[key] = fields.difference(non_upsert_fields.get(key)) 
+            else:
+                actual_expected[key] = fields
+        if stream:
```

Comment:
Looks like this goes through the whole dictionary and generates actual_expected for each stream but then only sends one stream if we passed in a stream otherwise send everything.  Might be more efficient to check if we sent stream before doing all the work.  Then we can just do the difference for one stream and return or do the whole dict and return.  Probably not a big deal either way but may be nice to have.

Reply:
Changed this as suggested